### ISSUE TITLE
Beta branch ignores published releases on master

### ISSUE THREAD

**Issue by AndriyGo** [semantic-release/semantic-release#1603]
## Current behavior

I created `beta` branch from `master` when the last published release was 1.1.1. Then:

Add bug fix to beta – pre-release published as 1.1.2-beta.1 (GOOD)
Add another bug fix to beta – pre-release published as 1.1.2-beta.2 (GOOD)
Create and merge PR into master – release created for 1.1.2 (GOOD)
Add another bug fix to beta – pre-release published as 1.1.2-beta.3 (WRONG)

## Expected behavior

At the last step, I was expecting to see a pre-release of 1.1.3-beta.1

## Environment

- **semantic-release** version: 17.1.1 (running through GitHub action)
- CI environment: GitHub actions
- **semantic-release** configuration: below is a portion of my package.json file:

```
"release": {
    "plugins": [
      "@semantic-release/commit-analyzer",
      "@semantic-release/release-notes-generator",
      "@semantic-release/github"
    ],
    "branches": ["master", {"name": "beta", "prerelease": true}]
  }
```

**Comment by travi** [comment]
semantic-release uses the underlying behaviors of git, so the branching model applies in this case. your beta branch knows nothing about changes to your mainline that were made after the branch was created. in order to make those changes available to that branch, and therefore to semantic-release acting on that branch, you need to merge you mainline branch into the beta branch.

**Comment by AndriyGo** [comment]
@travi thank you for your response.

I am somewhat puzzled by this. In our workflow, there is no real need to merge anything back. And it kind of feels like an overkill.

Looking at the code, it seems that the reason for this behaviour is that `getTags` function limits results to the current branch. But why? Why not use `git tag` directly to get all tags?

Also, the docs mention that a project can have up to 3 release branches. If the situation for them is as described above, then there can be 3 different releases having the same release number.

Is there any good reason for this behaviour, or would it be better if it got fixed?

**Comment by travi** [comment]
i believe the primary reason for this behavior is that publishing `1.1.3-beta.1` assumes that all that was in `1.1.2` is included. that is common for versioning, regardless of which tool you use, like semantic-release.

admittedly, that doesnt prevent a version being published that is now behind what was published to `@latest`, but the solution to that is to merge mainline into your pre-release branch, as i mentioned before. that brings the missing changes into your branch, enabling the proper version to be published including the relevant changes. even if you mistakenly published the incorrect version before merging, at worst you can mark that one as deprecated and follow up by merging the mainline and publishing again.

there are some additional safe guards in place to prevent actual collisions (also, npm prevents publishing a version more than once) that i wouldn't articulate well enough off the top of my head, but keeping your pre-release branches up to date with your mainline is the best way to keep things in good standing.

**Comment by AndriyGo** [comment]
@travi that does make sense. However, I still feel that in some workflows an option to scan all repos for tags would be welcome.

Take our case as an example. We are building mobile apps and they can go through several stages of testing. For both iOS and Android there is almost an unlimited number of test tracks/tester groups you can create. As an example, let's assume that we want to go through these stages:

1) Internal test by team members (branch `alpha`)
2) Testing by testers whom we hired (branch `test`)
3) Public beta test by users who want latest version (branch `beta`)
4) Public release (branch `master`)

As you can see, all our builds can only progress from one branch to the next through PRs. We do not allow direct commits to branches apart from alpha. Now let's imagine that we pushed some changes to `alpha` and we are happy with them and want them to progress to `test`:

1) We make PR from `alpha` to `test`. When it is approved, we need to merge changes back to `alpha` for semantic release to work.
2) If we are happy with these changes after our testers review them, we create a PR from `test` to `beta`. If it is approved, changes need to be merged back from `beta` to `test`, and from `test` to `alpha`.
..... AND SO ON...

You can clearly see that in our case, such workflow makes no sense. If we enforce pipeline-like progression of our builds – merging changes back is a waste of time which, at a minimum, makes automation of our PRs extremely challenging, because some PRs to `alpha` require semantic-release + builds while others don't.

With our setup, we do not care that preceding branches fall behind upstream branches due to "merge PR" commits. Those will never affect our ability to merge future changes, so we don't care about them.

After your explanation, I do understand why the behaviour we want is unreasonable by default. However, I do not think that our setup is terrible – hence I believe that at a minimum there should be a way to configure it to work as we expect.

Maybe a flag to take into account all tags in all branches would do the job?

**Comment by travi** [comment]
i get what you are saying, but would suggest that you are dismissing an important part of the reality of the process. you cannot merge your pre-release line into the mainline without the changes that are missing from your pre-release branch from getting introduced. if you dont merge them into your branch during your testing, you will end up with a version released on `@latest` that includes changes that were untested (together at least) during your pre-release testing.

beyond this, with the process you are describing, i'm surprised that you are concerned about the accuracy of the semantic version. semantic versions are meant to communicate the level of change from the previous version. what you are asking for would purposely lie about the features of that previous version. i can understand wanting to follow a process like you describe, but in that type of a flow, the version numbers are not semantic and, in a way, mostly meaningless (at least until you do merge to mainline).

the way i would describe what you are asking for is that it is an abuse of the pre-releases functionality of semantic-release. you are essentially promoting a single version through various levels of pre-release maturity in the same way you might promote an application through various levels of testing environments. that is not the goal of semantic-release. the pre-releases functionality is intended to help with publishing releases in accordance with semver based on the changes from the previous version. if you are not including the changes from your mainline in that process, your flow simply does not align with the goals of semantic-release and you may need to leverage a different tool.

**Comment by AndriyGo** [comment]
@travi given our workflow, it is impossible to get changes in the `@latest` which were not tested before in the `beta`, so there is no real need to merge anything. The difference between them will always be due to "merge PR" commits. However, although those are, technically, "differences" – we can always be sure that they do not require any extra testing.

Maybe I did not describe our workflow precisely – each pre-release/release branch has three important characteristics:

1) Direct commits are prohibited
2) PRs IN are allowed ONLY from the previous staging branch
3) PRs OUT are allowed ONLY towards next staging branch.

Schematically, if we assume that our latest code is maintained on `develop` branch then it would look like this:

1) `develop` can be modified through different means (direct commits, PRs etc)
2) once we want to test something, PRs are triggered in this order: `develop` -> `alpha (internal testing)` -> `betaTesterGroup1` -> `betaTesterGroup2` -> ... ->`betaTesterGroupN` -> `master`.

Given this flow, there is never a real need to merge something back. Sure, branches will develop differences from `master` due to merge commits, but those can be safely ignored.

I do understand that this may not be the most standard workflow, but it does work quite well for mobile apps. Hence, I believe it would be great if there was an option to either ask semantic-release to scan all branches for tags or to specify the last release version manually. As I understand, if implemented, this functionality would need to be implemented in `commitAnalyzer`, so I will probably open a discussion there as well about this idea.

**Comment by travi** [comment]
while i understand that this type of a workflow may work well for your team, it is not likely to be a flow supported by this tool. the piece that i am trying to explain is that semantic-release is not a general workflow automation tool. it is a tool to facilitate properly publishing versions of your package in a way that follows semantic-versioning properly. 

your point that releasing a pre-release for a version that has already been published to `@latest` is accurate, but as i mentioned before, the proper way to resolve that issue in a way that follows semantic-versioning is to merge the mainline commits into your pre-release branch. in addition, because of properly following semantic-versioning, this enables your consumers to verify that the new features work as expected with the changes from the mainline that would otherwise be unavailable to verify in the pre-release version.

besides the complexity that would be added to this tool to support what you are proposing, it would promote versioning packages in a way that does not follow semantic-versioning. please recognize that this collection of tools is maintained for free in our spare time, so we have to be careful both that new functionality aligns with the goals and purpose of the project, but also that it does not add unnecessary maintenance burden. while the flow you describe may work ok for your team, it does not align with the goals of semantic-release, even optionally.

what i'm maybe struggling with is the semantics of the version you are proposing. since the version would be accurate once you merge your branch into the mainline, it might help to understand what you are thinking is communicated incorrectly by the version that is published. as you mentioned, it is possible to publish multiple pre-release lines, which means the predictive versioning of a pre-release is never guaranteed to match the version once it is actually promoted to `@latest`. that can only officially be determined once the branch is merged to mainline, regardless of steps taken in the branch. in terms of semantics, merging mainline to the branch enables publishing a version with the proper version number, including the changes suggested by that number, as a pre-release rather than those changes only being accounted for in the merge back to mainline. the only semantics i can think of that you might be trying to communicate with your request is the version that the new functionality would be published under, which can't be guaranteed anyway because other changes to mainline could impact that before this pre-release is merged.

finally, i mentioned it above, but to clarify further more reasoning why the flow you describe is not a fit for a tool encouraging proper semantic-versioning, the associated conventional pre-release names you are using for the stages of your flow are intended to communicate something different than you are using them for. as i mentioned before, what you describe is closer to promoting changes through testing environments than pre-releases that communicate stability or uncertainty. it sounds like you are essentially freezing a version and promoting based on human approvals rather than actual code changes. promotion from alpha to beta, for example should communicate that the api change being proposed has changed from very unstable or uncertain to a state of more stability or less uncertainty. it is accounting for changes being made during the development of that api, not just verifying that testing happened to verify the quality of the implementation.

i hope some of this reasoning helps understand where the behavior of this project comes from. if you are truly using semantic-release to align to semantic-versioning, my recommendation is to adjust your workflow to account for changes to be applied during the promotion process, which could include merging changes that have made their way into mainline. if your goal is instead to use pre-releases for the unintended purpose of supporting promotion of a frozen version through the stages of your testing process, either accept that it is an unintended use and that it comes with some details that might not align with your goals, or realize that semantic-release might not be the appropriate tool to achieve your goal.

**Comment by AndriyGo** [comment]
@travi thank you, some of your points are extremely interesting, especially regarding the semantic meaning of `alpha`, `beta` etc.

I guess the main reasons why we want to utilise semantic-release (especially in conjunction with commitizen) are:

1) Standardised commit messages;
2) Automatic versioning based on code changes.

These work well even in our current flow (releases are published with correct versions and correct changelogs even though pre-releases are off).

At the same time, we also want to utilise CI/CD which automates our testing. It is really the automation which makes pre-releases from different branches so appealing – we can trigger our builds based on tags generated by semantic-release and inform our testers about changes from the previous alpha/beta version.

The reason why I am not so eager to create back merges is that it may sometimes be convenient to push changes directly to `alpha`. However, if `alpha` got merged with "changes" from `master`, then push will be rejected. It is certainly something that can be dealt with, but a bit inconvenient.

An alternative could be automatic back merges from `master` to `beta` and `alpha`. In this case, PRs from `develop` to `alpha` could still work as we want, and pre-releases would have a correct version.

While doable, I think that this approach is a little bit more tedious, as the automation becomes more complex.

A better solution could be some custom code injection into semantic-release, possibly through the plugin. I do realise that we could fork commit-analyzer and implement changes which we need. However, I do not like the idea of a custom implementation which prevents us from receiving the latest updates to commit-analyzer.

Would it be possible to discuss some possibility of extending commit-analyzer to let developers run custom code before tags are created?

**Comment by ccrowley96** [comment]
I'm running into this exact same issue.

My team is setting up and open source library with the following development flow.

**Branches**

-  `main` - pre-release branch. All dev PRs go into this branch.  When new changes are merged / pushed, the @beta npm channel is published allowing anyone who consumes the library to subscribe to @beta via pre-release.

- `release` - Once a feature set is stable is tested in beta, we open a PR from `main` into `release`.  This publishes to the @latest npm channel.

It feels redundant and error prone to merge `release` back into `main` after each release PR, just to make sure the tags are updated.  If both `main` and `release` are protected branches, creating a release would mean: 

1. Open PR into `release` from `main`
2. Merge the PR (pushed to @latest)
3. Open PR from `release` into `main`
4. Merge the PR to update tags resetting @beta to latest release version

**Comment by moroine** [comment]
Currently I manage to get things work with git workflow (master, develop, feature branch) with the following config:

```
const mainConfig = {
  branches: [
    'master',
    { name: 'develop', channel: 'beta', prerelease: 'beta' },
  ],
  repositoryUrl: 'https://github.com/nugit/metadata-service.git',
  plugins: [
    '@semantic-release/commit-analyzer',
    '@semantic-release/release-notes-generator',
    [
      '@semantic-release/changelog',
      {
        changelogFile: 'CHANGELOG.md',
      },
    ],
    '@semantic-release/npm',
    [
      '@semantic-release/git',
      {
        assets: ['CHANGELOG.md'],
        message:
          // eslint-disable-next-line no-template-curly-in-string
          'chore(release): ${nextRelease.version} [skip ci]',
      },
    ],
    '@semantic-release/github',
    [
      '@semantic-release/exec',
      {
        publishCmd: 'git checkout -- package.json',
      },
    ],
    [
      'semantic-release-slack-bot',
      {
        notifyOnSuccess: true,
        notifyOnFail: true,
        markdownReleaseNotes: true,
      },
    ],
  ],
};

const { execSync } = require('child_process');
const { createHash } = require('crypto');

const branch = execSync('git branch --show-current').toString().trimEnd('\n');
const channel = createHash('md5').update(branch).digest('hex');

const localConfig = {
  branches: [
    'master',
    { name: 'develop', channel: 'beta', prerelease: 'beta' },
    {
      name: branch,
      channel,
      prerelease: channel,
    },
  ],
  repositoryUrl: 'git@github.com:nugit/metadata-service.git',
  plugins: [
    '@semantic-release/commit-analyzer',
    '@semantic-release/npm',
  ],
};

module.exports = process.env.LOCAL_RELEASE ? localConfig : mainConfig;
```

You can see that I have two configs, one for the workflow and another which allows me to publish pre-release of the feature branch without conflicting with other versions and generating changelog

**Comment by inabhi9** [comment]
Our workflow is also around (release/development/feature) branches as other mentioned, and don't see any need to merge back release branch to development branch manually.

Since, we develop end application but library, we aren't using npm packages for publish; ideal semantic-versioning flow doesn't make much sense in our flow.

I would also like to have some sort of option to define a branch to scan tags from. For now, this is what I come up with,

```js
// git-tag-monkeypatch.js

const git = require('semantic-release/lib/git');

const _gitTags = git.getTags;

git.getTags = async (branch, execaOptions) => {
    const branchObj = config.branches.find(b => b.name === branch);
    branch = branchObj.tagsFrom || branch;

    return _gitTags(branch, execaOptions);
}
```

and config looks like

```js
module.exports = {
    "branches": [
        {
            "name": "main"
        },
        {
            name: "development",
            prerelease: "beta",
            tagsFrom: 'main'
        },
        {
            "name": "staging",
            "prerelease": "rc",
            "tagsFrom": "development"
        }
    ]
};
```

and have this command in npm scripts
```json
{
  {
    "semantic-release": "node -r ./git-tag-monkeypatch.js node_modules/.bin/semantic-version"
  }
}
```

**Comment by polpenaloza** [comment]
Solution: rebase all your branches with production. Not ideal, not great. Actually, pretty bad dev experience.

**Comment by tomavic** [comment]
> Solution: rebase all your branches with production. Not ideal, not great. Actually, pretty bad dev experience.



Hi folks, 

I facing same issues and I have been into this for a long time. I have example app to demonstrate different scenarios, please check from here https://github.com/tomavic/myreads-tracker-nx-react-semantic/pull/17


When I tried to create pul request from my next to master and rebuild my main branch, I got this error:
<img width="769" alt="Screenshot 2023-04-22 at 3 22 21 PM" src="https://user-images.githubusercontent.com/17650005/233787625-3d172931-6e3f-4c34-a7a8-75f81dc3a23c.png">


Please check the failed release from here: https://github.com/tomavic/myreads-tracker-nx-react-semantic/actions/runs/4772970879/jobs/8485769937

**Comment by travi** [comment]
@tomavic please open a separate issue. Your symptoms are not related to the topic of this thread

**Comment by tomavic** [comment]
@travi we don't need a separate issue here, this topic has so many issues to be tracked and they are all related. There are so many git flows and different setup for each one.

**Comment by lainas262** [comment]
Hey everyone, I followed the conversation in this issue as I noticed the behavior in regards to beta release version bumps after a release on the mainline. I understand the flow and reasoning behind having to merge mainline back to beta but I am a bit confused on how to handle merge conflicts. 

Not code conflicts of course :laughing: but conflicts with auto-generated changelog file and the conflict with versions on package.json since the two branches have diverged. Is there a process to follow? 

For example in my official release changelog from the mainline branch, should I include intermediate logs from beta and simply merge both sides? This doesn't make sense to me as this could end up with duplicate release messages on the mainline changelog - entries for all beta releases in between, as well as the main entry of the next version which will include in a batch everything that previous beta releases included. 

Alternatively, should I just always keep the changes from the destination branch when resolving conflicts? So mainline branch will always have @latest release logs in changelog, beta branch will always have @beta release logs in changelog etc? 

I tested the first approach and changelog seems to have a lot of duplication. I will run some tests to check the behavior of the latter. To see what will happen between merges of mainline to beta and beta to mainline to see the result of the logs when always keeping changelog file and package.json file as it was from the destination branch.

Any advise would be helpful. Also apologies if this is posted at the wrong place, let me know if I should move it under the specific plugin's repo
https://github.com/semantic-release/changelog

**Comment by lainas262** [comment]
Hmm actually now that I think about it. Maybe merging from both isn't such a bad idea when it comes to changelog changes? because this will include each change for each version ever published regardless of distribution channel. So I guess having everything together on the mainline branch isn't such a problem? I am trying to understand what's more "semantically" correct in terms of workflow
